compiler/rustc_mir_transform/src/simplify_pow_of_two.rs

```diff
@@ -0,0 +1,268 @@
+//! A pass that checks for and simplifies calls to `pow` where the receiver is a power of
+//! two. This can be done with `<<` instead.
+
+use crate::MirPass;
+use rustc_const_eval::interpret::{ConstValue, Scalar};
+use rustc_hir::definitions::{DefPathData, DisambiguatedDefPathData};
+use rustc_middle::mir::patch::MirPatch;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt, UintTy};
+use rustc_span::sym;
+use rustc_target::abi::FieldIdx;
+
+pub struct SimplifyPowOfTwo;
+
+impl<'tcx> MirPass<'tcx> for SimplifyPowOfTwo {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let mut patch = MirPatch::new(body);
+
+        for (i, bb) in body.basic_blocks.iter_enumerated() {
+            let term = bb.terminator();
+            let source_info = term.source_info;
+            let span = source_info.span;
+
+            if let TerminatorKind::Call {
+                    func,
+                    args,
+                    destination,
+                    target: Some(target),
+                    unwind,
+                    call_source: CallSource::Normal,
+                    ..
+                } = &term.kind
+                && let Some(def_id) = func.const_fn_def().map(|def| def.0)
+                && let def_path = tcx.def_path(def_id)
+                && tcx.crate_name(def_path.krate) == sym::core
+                // FIXME(Centri3): I feel like we should do this differently...
```

Comment:
Perhaps check that it's an associated method of an integer primitive type?

Using names is usually the wrong way to identify things.

(Arguably maybe it should be a lang item if it's recognized semantically, but that would probably need one lang item per type, which would be a mess...)

Reply:
Lang items would make sense, though yeah, there'd be tons of them... I tried with diagnostic items before and it was a bit of pain due to that.

Reply:
InstSimplify looks at the declaring type for one of its things, so maybe you can use something like that?

https://github.com/rust-lang/rust/blob/89acdae9f243da15549c7a8bc737e6240c301790/compiler/rustc_mir_transform/src/instsimplify.rs#L203-L219

Reply:
We'd still need to make sure the method name is `pow`, though that should be a lot better.

Reply:
I think the most graceful way to do this is to add a diagnostic item: https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-items.html

Reply:
Diagnostic items should never be used for semantic things like this.  If it affects the output of rustc, it's a lang item, not a diagnostic item.

Reply:
So... effectively what would be ideal would be some sort of `is_known_to_be_const` intrinsic that can be used on runtime values to switch to a more optimal path (and could spuriously return `false` if it is too hard to figure out). Then, if `pow` gets inlined (as only then we'd be able to answer this with `true`), only the optimized path would stay around.

To make this work for functions that are too big to get inlined, it may be necessary to create another MIR optimization that copies functions that get passed constants as arguments (and that don't get inlined), removes those constant arguments from the copy, and replace the arguments inside the copy's body with these constants. GCC does this optimization at opt-level 3, and LLVM probably does it, too.

But `pow` is currently not getting inlined by MIR inlining. Maybe we could add some `rustc_*` flag that inlines a function if any of its arguments are constant or sth.

All of this is a lot of work, too, but I believe they would allow `pow` to stay `const fn`, while also avoiding adding 10 new lang items that are all effectively the same. And the scheme would probably enable other similar optimizations.

Reply:
> but I believe they would allow `pow` to stay `const fn`

Does `rustc_lang_item` require it to not be const?

Reply:
> Does `lang` require it to not be const?

No, but we could have a single generic lang item that uses trait bounds for `Add`, `Mul` and such. Then all the integer impls could call the generic lang item, and we could make our optimization work on the generic lang item. Though that would still have to happen post-inlining

Reply:
LLVM has an llvm.is.constant intrinsic, which basically does what you want is_known_to_be_const to do. If you export that intrinsic and add a code path for handling power-of-two guarded by it to the pow implementation, that should get optimized fine on the LLVM level.

Reply:
That would still get the power used and all at runtime though, afaict, as you can't use a non-constant in a `const` block (which would be `exp` here). I believe this requires at least a `pow` call anyway so if it does it at runtime it'll be slower regardless.

Reply:
I think the trick is that if you add a

```rust
if llvm_is_constant(self) {
    if self.is_power_of_two() {
         ...
    }
}
```

then, since LLVM knows `self` is a constant, it will be able to optimize `self.is_power_of_two()` into `true` or `false`, and then either optimize out your logic or the rest of `pow` depending on the result.

Reply:
Good news: I did that and the generated code is *even better* with it, and it 100% works! Will open a separate PR for that and close this one (though not rn as the intrinsic will cause rustc to segfault if called on non-i32s, will need some more work :upside_down_face:)